File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/TypedAssetSerializator.java
package com.greentree.engine.moon.assets.serializator;

import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.commons.reflection.info.TypeInfoBuilder;
import com.greentree.commons.reflection.info.TypeUtil;

public abstract class TypedAssetSerializator<T> implements AssetSerializator<T> {

    public final TypeInfo<T> TYPE;

    protected TypedAssetSerializator() {
        TYPE = TypeUtil.getFirstAtgument(TypeInfoBuilder.getTypeInfo(getClass()),
                TypedAssetSerializator.class);
    }

    public TypedAssetSerializator(Class<T> cls) {
        this(TypeInfoBuilder.getTypeInfo(cls));
    }

    public TypedAssetSerializator(TypeInfo<T> type) {
        TYPE = type;
    }

    @Override
    public TypeInfo<T> getType() {
        return TYPE;
    }

}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/TypedAssetSerializator.kt
package com.greentree.engine.moon.assets.serializator

import com.greentree.commons.reflection.info.TypeInfoBuilder

abstract class TypedAssetSerializator<T : Any>() : AssetSerializator<T> {

	constructor(cls: Class<T>?) : this(TypeInfoBuilder.getTypeInfo<T>(cls))
}